### PR TITLE
Support external token source for flyteadmin clients

### PR DESCRIPTION
**Pull request by ckiosidis**
This PR aims to add support for external token source generation and usage when communicating with flyteadmin.

As discussed, I started by refactoring to introduce a TokenSourceProvider interface and the implementations for pkce and client credentials.

Looking for comments for the implementation before moving on and introducing an external_process AuthType where the TokenSourceProvider will run a shell command to generate a token/tokenSource.

Running external auth example:

```

  authType: ExternalProcess
  command: [gcloud,auth,print-identity-token]
```

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_
Refactored the token source, and introduce a TokenSourceProvider.
Then introduced an external process authentication type, that will execute a command passed from the config.
The command is expected in the format of a list of strings.
The result will be populated in the outgoing rpc towards the flyteadmin host.

## Tracking Issue
https://github.com/flyteorg/flyte/issues/1557

## Follow-up issue
_NA_

### CODE REVIEW

File: clients/go/admin/token_source_provider.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import (
+	"context"
+	"fmt"
+	"github.com/flyteorg/flyteidl/clients/go/admin/clientcredentials"
+	"github.com/flyteorg/flyteidl/clients/go/admin/pkce"
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/service"
+	"golang.org/x/oauth2"
+)
+
+type TokenSourceProvider interface {
```
Comment (kumare3) [inline clients/go/admin/token_source_provider.go:12]: documentation for the interface

File: clients/go/admin/token_source_provider.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+	GetTokenSource(ctx context.Context) (oauth2.TokenSource, error)
+	GetAuthType() AuthType
```
Comment (kumare3) [inline clients/go/admin/token_source_provider.go:14]: do we need this method?

**Review by kumare3 — COMMENTED** [review]
a small nit. But this looks good to me.

**Comment by pmahindrakar-oss** [comment]
@ckiosidis  can you test flytectl authentication also using these changes and see if flytectl is able to communicate propely using pkce type

File: clients/go/admin/token_source_provider.go (hunk 2 above)
Comment (ckiosidis) [reply clients/go/admin/token_source_provider.go:14]: no i can remove

**Comment by pmahindrakar-oss** [comment]
I tested with these changes in flytectl and they are working fine.

Replaced the flyteidl
```
go mod edit -replace github.com/flyteorg/flyteidl=github.com/ckiosidis/flyteidl@60e4494f82fb3dc2b024ff58db9073ac07f16c8f
```
Ran go mod 
```
go mod tidy
```
Compiled the flytectl binary
```
make compile
```

called get preojects to verify auth flow
```
flytectl get projects --logger.level=6
{"json":{"src":"auth_flow_orchestrator.go:37"},"level":"debug","msg":"got a response from the refresh grant for old expiry 2021-09-27 16:09:54.424807 +0530 IST with new expiry 2021-09-27 16:09:54.424807 +0530 IST","ts":"2021-09-27T15:42:19+05:30"}
{"json":{"src":"client.go:59"},"level":"info","msg":"Initialized Admin client","ts":"2021-09-27T15:42:19+05:30"}
```

**Review by pmahindrakar-oss — APPROVED** [review]

**Review by kumare3 — APPROVED** [review]

**Comment by EngHabu** [comment]
Will review shortly today.. till then, if there is no GitHub issue for this. Do you mind filling one and linking it here?

**Comment by pmahindrakar-oss** [comment]
@EngHabu here you go https://github.com/flyteorg/flyte/issues/1557 cc :  @ckiosidis

File: clients/go/admin/token_source_provider.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+			return nil, err
+		}
+	} else if cfg.AuthType == AuthTypeExternalProcess {
+		tokenProvider, err = NewExternalTokenSourceProvider(cfg.ExternalCommand)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		return nil, fmt.Errorf("unsupported type %v", cfg.AuthType)
+	}
```
Comment (EngHabu) [inline clients/go/admin/token_source_provider.go:58]: nit: I would refactor to a `switch`

File: clients/go/admin/token_source_provider.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+			return nil, fmt.Errorf("failed to fetch auth metadata. Error: %v", err)
+		}
+
+		tokenURL = metadata.TokenEndpoint
+	}
+
+	clientMetadata, err := authClient.GetPublicClientConfig(ctx, &service.PublicClientAuthConfigRequest{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch client metadata. Error: %v", err)
+	}
```
Comment (EngHabu) [inline clients/go/admin/token_source_provider.go:38]: I think you can safely move this into `AuthTypeClientSecret` case... avoid the extra call when not needed...

File: clients/go/admin/client.go [hunk 5]
```diff
@@ -84,51 +86,32 @@ func GetAdditionalAdminClientConfigOptions(cfg *Config) []grpc.DialOption {
...
 	}
 
 	oauthTokenSource := NewCustomHeaderTokenSource(tSource, cfg.UseInsecureConnection, clientMetadata.AuthorizationMetadataKey)
 	return grpc.WithPerRPCCredentials(oauthTokenSource), nil
 }
 
-// Returns the client credentials token source to be used eg by flytepropeller to communicate with admin/ by CI
-func getClientCredentialsTokenSource(ctx context.Context, cfg *Config,
-	clientMetadata *service.PublicClientAuthConfigResponse, tokenURL string) (oauth2.TokenSource, error) {
+type ClientCredentialsTokenSourceProvider struct {
```
Comment (EngHabu) [inline clients/go/admin/client.go:109]: nit: Maybe move this type along side the other token source providers?

File: clients/go/admin/config.go [hunk 6]
```diff
@@ -62,6 +64,8 @@ type Config struct {
 	DeprecatedAuthorizationHeader string `json:"authorizationHeader" pflag:",Custom metadata header to pass JWT"`
 
 	PkceConfig pkce.Config `json:"pkceConfig" pflag:",Config for Pkce authentication flow."`
+
+	ExternalCommand []string `json:"command" pflag:",Command for external authentication token generation"`
```
Comment (EngHabu) [inline clients/go/admin/config.go:68]: nit: there is a convention that json tags match the name of the field... so I would either rename the field to `Command` or the json tag to `externalCommand`... I'm ok with the former

File: clients/go/admin/config.go [hunk 7]
```diff
@@ -32,6 +32,8 @@ const (
 	AuthTypeClientSecret AuthType = iota
 	// Chooses Proof Key Code Exchange OAuth2 extension protocol (ref: https://tools.ietf.org/html/rfc7636)
 	AuthTypePkce
+	// Chooses an external authentication process
+	AuthTypeExternalProcess
```
Comment (EngHabu) [inline clients/go/admin/config.go:36]: nit: rename to match the name of the field to fill in? just for convenience...

```suggestion
	AuthTypeExternalCommand
```

**Review by EngHabu — COMMENTED** [review]
Hey @ckiosidis thank you for making this code better and more generic! I have a quick few nits if you can take a look, would be awesome!

**Comment by ckiosidis** [comment]
@EngHabu fixed the all the comments, please take a look.
I rebased to sign-off and had to force push

**Review by pmahindrakar-oss — APPROVED** [review]

**Review by EngHabu — APPROVED** [review]
Thank you!

**Pull request merged**